### ISSUE TITLE
PixelShuffle is very slow on MPS compared to ConvTranspose2d and PixelShuffle on cuda

### ISSUE THREAD

**Issue by neil3706** [pytorch/pytorch#83196]
### 🐛 Describe the bug

I changed the decode layers of a CNN to use pixel shuffle instead of transpose_conv2d. The model immediately took 3x longer to run. GPU usage was still high and there was no warning about fallback to using the CPU. I recently starting training the CNN remotely using coda and saw a reduction in training by 66%. The only conclusion is that there is a problem with the MPS implementation.

The commented out lines show the ConvTranspose2D  being replaced with the Conv2d followed by PixelShuffle

class decodeUpBlock(nn.Module):
    def __init__(self, inFM, outFM, kSize, outSize, stride=1, padding=0, output_padding=0):
        super(decodeUpBlock, self).__init__()
        self.layers = nn.Sequential()
        #self.layers.add_module("TConv", nn.ConvTranspose2d(inFM,outFM,kSize,stride=stride,padding=padding,output_padding=output_padding))
        self.layers.add_module("Conv2", nn.Conv2d(inFM,4*outFM,kSize,stride=1,padding=padding))
        self.layers.add_module("pShfl", nn.PixelShuffle(stride))
        self.layers.add_module("lNorm", nn.LayerNorm([outFM,outSize,outSize]))
        self.layers.add_module("ReLU", nn.LeakyReLU(0.2,inplace=True))

### Versions

Collecting environment information...
PyTorch version: 1.12.0.post2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:07:06) [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] torch==1.12.0.post2
[pip3] torchvision==0.13.0a0
[conda] numpy                     1.23.1          py310h0a343b5_0    conda-forge
[conda] pytorch                   1.12.0          cpu_py310h911b1ea_2    conda-forge
[conda] torchvision               0.13.0          cpu_py310he68663e_0    conda-forge

**Comment by neil3706** [comment]
updated to latest nightly build on 8/10 with no change. Tried implementing the function using reshape and permute and there wasn't any improvement to the performance.

def shuffle(x, r):
    [B, C, H, W] = list(x.size())
    x = x.reshape(B, C//(r*r),r,r, H,W)
    x = x.permute(0, 1, 4, 2, 5, 3)
    x = x.reshape(B, -1, H*r, W*r)
    return x

**Comment by kulinseth** [comment]
@DenisVieriu97 , please take a look.

**Comment by alexdremov** [comment]
@kulinseth, @razarmehr 

Are there any updates on this? The problem is reproduced in 2.1.0 nightly build. MPS implementation is extremely slow. I tried several methods:

```python
import torch.nn.functional as F

class PixelShuffle(nn.Module):
    def __init__(self, upscale_factor):
        super(PixelShuffle, self).__init__()
        self.upscale_factor = upscale_factor

    def forward(self, x):
        batch_size, channels, height, width = x.size()

        # Calculate the new height and width
        new_height = height * self.upscale_factor
        new_width = width * self.upscale_factor

        # Calculate the channel size after shuffling
        new_channels = channels // (self.upscale_factor ** 2)

        # Reshape the input tensor
        x = x.view(batch_size, new_channels, self.upscale_factor, self.upscale_factor, height, width)

        # Transpose and flatten the tensor
        x = x.permute(0, 1, 4, 2, 5, 3).contiguous()
        x = x.view(batch_size, new_channels, new_height, new_width)

        return x
```

This is similar to the [neil3706's](https://github.com/neil3706) code.

I also tried version that uses unfold, but it fallbacks to CPU and, obviously, results in low performance:

```python
import torch.nn.functional as F

class PixelShuffle(nn.Module):
    def __init__(self, upscale_factor):
        super(PixelShuffle, self).__init__()
        self.upscale_factor = upscale_factor

    def forward(self, x):
        batch_size, channels, height, width = x.size()

        # Calculate the new height and width
        new_height = height * self.upscale_factor
        new_width = width * self.upscale_factor

        # Calculate the channel size after shuffling
        new_channels = channels // (self.upscale_factor ** 2)

        # Reshape the input tensor using unfold()
        x = F.unfold(x, kernel_size=self.upscale_factor, stride=self.upscale_factor)
        x = x.view(batch_size, new_channels, self.upscale_factor, self.upscale_factor, height, width)
        x = x.permute(0, 1, 4, 2, 5, 3).contiguous()
        x = x.view(batch_size, new_channels, new_height, new_width)

        return x
```
